packages/server/src/main/java/software/uncharted/terarium/hmiserver/models/TerariumEntity.java

```diff
@@ -0,0 +1,63 @@
+package software.uncharted.terarium.hmiserver.models;
+
+import io.swagger.v3.oas.annotations.media.Schema;
+import jakarta.persistence.Column;
+import jakarta.persistence.Id;
+import jakarta.persistence.MappedSuperclass;
+import jakarta.persistence.PrePersist;
+import jakarta.persistence.PreUpdate;
+import java.io.Serializable;
+import java.sql.Timestamp;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.UUID;
+import lombok.Data;
+import lombok.experimental.Accessors;
+import software.uncharted.terarium.hmiserver.annotations.TSModel;
+import software.uncharted.terarium.hmiserver.annotations.TSOptional;
+
+@Data
+@Accessors(chain = true)
+@TSModel
+@MappedSuperclass
+public abstract class TerariumEntity implements Serializable {
+
+	@Id
+	@TSOptional
+	@Schema(accessMode = Schema.AccessMode.READ_ONLY)
+	private UUID id = UUID.randomUUID();
+
+	@TSOptional
+	@Schema(accessMode = Schema.AccessMode.READ_ONLY)
+	@Column(columnDefinition = "TIMESTAMP WITH TIME ZONE")
+	private Timestamp createdOn;
+
+	// Don't use @CreationTimestamp because it doesn't get populated until after the
+	// transaction is committed.
+	@PrePersist
+	protected void onCreate() {
+		this.createdOn =
+				Timestamp.from(ZonedDateTime.now(ZoneId.systemDefault()).toInstant());
+		this.updatedOn = this.createdOn;
+	}
+
+	@TSOptional
+	@Schema(accessMode = Schema.AccessMode.READ_ONLY)
+	@Column(columnDefinition = "TIMESTAMP WITH TIME ZONE")
+	private Timestamp updatedOn;
```

Comment:
is `deletedOn` not here?

Reply:
It's on TerariumAsset, I am now realizing we will need to set up triggers to cascade our soft deletes....

Reply:
Right I saw it there - I'm just curious why `deletedOn` wasn't moved down to `TerariumEntity` as well

Reply:
I left it out since that logic is built into the TerariumAssetService stuff, and those nested fields don't have their own services.